Sources/SPTDataLoaderSwift/Request.swift

```diff
@@ -0,0 +1,287 @@
+ specific language governing permissions and limitations
+ under the License.
+ */
+
+import Foundation
+
+typealias ResponseHandler = (Result<SPTDataLoaderResponse, Error>) -> Void
+public typealias ResponseValidator = (SPTDataLoaderResponse) throws -> Void
+
+/// A wrapper for initiating URL requests and handling responses with optional validation.
+///
+/// The `Request` type provides a set of functions that make it easy to chain a series of
+/// response handlers and validators.
+///
+/// The request is executed upon attachment of the first response handler. If a handler is
+/// attached after the response has been received, it will be immediately invoked with the
+/// existing value.
+public final class Request {
+    private let request: SPTDataLoaderRequest
+    private let executionHandler: () -> SPTDataLoaderCancellationToken?
+
+    init(request: SPTDataLoaderRequest, executionHandler: @escaping () -> SPTDataLoaderCancellationToken?) {
+        self.request = request
+        self.executionHandler = executionHandler
+    }
+
+    private enum State {
+        case initialized
+        case executed(token: SPTDataLoaderCancellationToken)
+        case completed(response: SPTDataLoaderResponse)
+        case failed(error: Error)
+        case cancelled
+    }
+
+    private let accessLock = AccessLock()
+    private var state: State = .initialized
+    private var responseHandlers: [ResponseHandler] = []
+    private var responseValidators: [ResponseValidator] = []
+
+    func addResponseValidator(_ responseValidator: @escaping ResponseValidator) {
+        accessLock.sync {
+            switch state {
+            case .initialized, .executed:
+                responseValidators.append(responseValidator)
+            default:
+                break
+            }
+        }
+    }
+
+    func addResponseHandler(_ responseHandler: @escaping ResponseHandler) {
+        var result: Result<SPTDataLoaderResponse, Error>?
+
+        accessLock.sync {
+            switch state {
+            case .initialized:
+                if let token = executionHandler() {
+                    responseHandlers.append(responseHandler)
+                    state = .executed(token: token)
+                } else {
+                    result = .failure(RequestError.executionFailed)
+                    state = .failed(error: RequestError.executionFailed)
+                }
+            case .executed:
+                responseHandlers.append(responseHandler)
+            case .failed(let error):
+                result = .failure(error)
+            case .completed(let response):
+                result = .success(response)
+            default:
+                break
+            }
+        }
+
+        result.map { result in responseHandler(result) }
+    }
+
+    func processResponse(_ response: SPTDataLoaderResponse) {
+        var error = response.error
+        var handlers: [ResponseHandler] = []
+
+        accessLock.sync {
```

Comment:
same here, you should only lock on things that are mutated that are owned by this object

Reply:
As above, I believe this whole section is critical because of the read + mutate access to `state`, `responseHandlers`, and `responseValidators `.